### PR TITLE
Adding test to validate worker invocation metrics

### LINKED ISSUE #7292: Recycle failing language workers
Have seen a few scenarios where the functions host is running multiple language workers and some of them are consistently failing, while others are working fine. Often this involves some user code component going into a bad state.

In most cases this is resolved with a restart on the function app, but this is a heavy handed approach that requires manual intervention.  The functions host should detect this scenario and attempt to resolve it by recycling the language workers that are failing. It will still be up to the user to analyze their logs and fix their problems, but the platform can reduce the impact until that happens by automatically recycling failing workers.

The single language worker case is a bit trickier. There's no obvious signal that a restart will fix the problem. I'd suggest the first version of this improvement should only kick in when there are multiple language workers.

### PR DESCRIPTION
**Pull request by surgupta-msft**
Added E2E test for the Metrics added here - [commit](https://github.com/Azure/azure-functions-host/commit/20fbae3fad9384d53ac57b8f020f71b50e07d250#diff-32fe7dff394f43a757279f5626eea4e39295fc6b5909c638a6934dac4aa42c6aR28)

### Issue describing the changes in this PR

resolves #7292 

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs [hunk 1]
```diff
@@ -350,9 +350,10 @@ public async Task SendInvocationRequest_PublishesOutboundEvent()
             ScriptInvocationContext scriptInvocationContext = GetTestScriptInvocationContext(Guid.NewGuid(), null);
             await _workerChannel.SendInvocationRequest(scriptInvocationContext);
             await Task.Delay(500);
+            string testWorkerId = _workerId.ToLowerInvariant();
             var traces = _logger.GetLogMessages();
             Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, _expectedLogMsg)));
-            Assert.False(string.IsNullOrEmpty(_metricsLogger.LoggedEvents.FirstOrDefault(e => e.Contains("testworkeridworker.invoke_js1"))));
+            Assert.False(string.IsNullOrEmpty(_metricsLogger.LoggedEvents.FirstOrDefault(e => e.Contains($"{string.Format(MetricEventNames.WorkerInvoked, testWorkerId)}_{scriptInvocationContext.FunctionMetadata.Name}"))));
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs]: Are we expecting 1 and only 1? If so we should assert the count, e.g. Assert.Equal(1, logger.LoggedEvents.Count(e => e == "...")

File: test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs [hunk 2]
```diff
@@ -688,6 +689,24 @@ public void SendFunctionEnvironmentReloadRequest_WithDirectory()
             Assert.True(envReloadRequest.FunctionAppDirectory == _scriptRootPath);
         }
 
+        [Fact]
+        public async Task SendInvocationRequest_PublishesOutboundEvent_ReceiveInvocationResponse()
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs]: nit: _ReceivesInvocationResponse

File: test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs [hunk 3]
```diff
@@ -688,6 +689,24 @@ public void SendFunctionEnvironmentReloadRequest_WithDirectory()
...
+            _metricsLogger.ClearCollections();
+            var invocationid = Guid.NewGuid();
+            ScriptInvocationContext scriptInvocationContext = GetTestScriptInvocationContext(invocationid, new TaskCompletionSource<ScriptInvocationResult>());
+            await _workerChannel.SendInvocationRequest(scriptInvocationContext);
+            _testFunctionRpcService.PublishInvocationResponseEvent(invocationid.ToString());
+            await Task.Delay(500);
+            var testWorkerId = _workerId.ToLowerInvariant();
+            var traces = _logger.GetLogMessages();
+            Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, $"InvocationResponse received for invocation id: '{invocationid}'")));
+            Assert.False(string.IsNullOrEmpty(_metricsLogger.LoggedEvents.FirstOrDefault(e => e.Contains($"{string.Format(MetricEventNames.WorkerInvoked, testWorkerId)}_{scriptInvocationContext.FunctionMetadata.Name}"))));
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs]: Same comment as above. We should assert the count of all logged metrics (3), then verify the exact metrics. E.g. just ToArray the LoggedEvents, check it's Length, then check each index for the metric you expect

File: test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs (hunk 1 above)
Comment (surgupta-msft) [reply test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs]: Done

File: test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs (hunk 3 above)
Comment (surgupta-msft) [reply test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs]: Updated code here

File: test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs (hunk 2 above)
Comment (surgupta-msft) [reply test/WebJobs.Script.Tests/Workers/Rpc/GrpcWorkerChannelTests.cs]: Done

**Review by mathewc — APPROVED** [review]

**Review by liliankasem — APPROVED** [review]

**Pull request merged**